keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,374 @@
+is set to a very high number to allow for disruptions (which occur often at a large scale like this), the workload could
+waste roughly `maxRestarts` number of minutes repeatedly recreating despite the fact that it is doomed to fail due to a non-retriable
+error like an application code bug. This is very expensive and wasteful on expensive, large scale clusters.
+
+We need to define a more configurable Failure Policy API which will allow users to define their own restart and failure semantics, 
+based on the type of child job failure occurring.
+
+### Goals
+
+- Enhance the Failure Policy API with options allowing the user to control the restart and failure semantics of the JobSet.
+
+### Non-Goals
+
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, in order to use my computing resources as efficiently as possible, I want to 
+configure my JobSet to restart in the event of a child job failure due to a retriable error
+like host maintance, but to fail the JobSet immediately without any unnecessary restart 
+attempts in the event of an non-retriable application code error. 
+
+**Example Failure Policy Configuration for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: configurable-failure-policy-1
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: cloud.google.com/gke-nodepool # 1:1 job replica to node pool assignment
+spec:
+  # Failure Policy configured to fail the JobSet immediately if a job failure reason for any child job was NOT due to a SIGTERM (e.g. host maintenance).
+  # Otherwise, restart up to 10 times.
+  failurePolicy:
+    rules:
+    - action: FailJobSet
+      onJobFailureReasons: 
+      - PodFailurePolicy
+    maxRestarts: 10
+  replicatedJobs:
+  - name: buggy-job 
+    replicas: 1 # set to number of node pools
+    template:
+      spec:
+        parallelism: 1 # number of nodes per pool
+        completions: 1 # number of nodes per pool
+        backoffLimit: 0
+        # podFailurePolicy which fails job immediately if job was not killed by SIGTERM (i.e., graceful node shutdown for maintenance events)
+        podFailurePolicy:
+          rules:
+          - action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: NotIn
+              values: [143] # SIGTERM = exit code 143
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: bash:latest
+              image: docker.io/library/bash:5
+              command: ["bash"]
+              args:
+              - -c
+              - echo "Hello world! I'm going to exit with exit code 1 to simulate a software bug." && sleep 20 && exit 1
+```
+
+### Story 2
+
+As a user, I have a JobSet with 2 replicated jobs: one which runs distributed training processes across a pool of GPU/TPU
```

Comment:
Again, these examples would be hard for someone to reproduce if they aren't using GKE. Are these details necessary for the KEP?

Reply:
Updated example to be more cloud neutral.